--- decoder_plugins/mp3/mp3.c	2007-03-22 20:32:34.000000000 +0100
+++ decoder_plugins/mp3/mp3.ch.c	2009-12-08 23:31:12.000000000 +0100
@@ -153,28 +153,7 @@
 	if (frame && (field = &frame->fields[1])) {
 		ucs4 = id3_field_getstrings (field, 0);
 		if (ucs4) {
-			
-			/* Workaround for ID3 tags v1/v1.1 where the encoding
-			 * is latin1. */
-			if (id3_tag_options(tag, 0, 0) & ID3_TAG_OPTION_ID3V1) {
-				char *t;
-
-				comm = (char *)id3_ucs4_latin1duplicate (ucs4);
-
-#ifdef HAVE_RCC
-				if (options_get_int("UseRCC"))
-					comm = do_rcc (comm);
-				else {
-#endif /* HAVE_RCC */
-					t = comm;
-					comm = id3v1_fix (comm);
-					free (t);
-#ifdef HAVE_RCC
-				}
-#endif /* HAVE_RCC */
-			}
-			else
-				comm = (char *)id3_ucs4_utf8duplicate (ucs4);
+			comm = (char *)id3_ucs4_utf8duplicate (ucs4);
 		}
 	}
 
